logstash-core/src/main/java/org/logstash/common/io/MmapPageIO.java

```diff
@@ -1,199 +1,74 @@
-    private final int capacity;
-    private final String dirPath;
-    private final int pageNum;
-    private final List<Integer> offsetMap; // has to be extendable
-
-    private MappedByteBuffer buffer;
+public class MmapPageIO extends AbstractByteBufferPageIO {
+
     private File file;
     private FileChannel channel;
-
-    private long minSeqNum; // TODO: to make minSeqNum final we have to pass in the minSeqNum in the constructor and not set it on first write
-    private int elementCount;
-    private int head;
-    private byte version;
-    private Checksum checkSummer;
+    protected MappedByteBuffer buffer;
 
     public MmapPageIO(int pageNum, int capacity, String dirPath) throws IOException {
-        this.pageNum = pageNum;
-        this.capacity = capacity;
-        this.dirPath = dirPath;
-        this.offsetMap = new ArrayList<>();
-        this.checkSummer = new CRC32();
+        super(pageNum, capacity);
+
+        this.file = Paths.get(dirPath, "page." + pageNum).toFile();
     }
 
     @Override
     public void open(long minSeqNum, int elementCount) throws IOException {
-        this.minSeqNum = minSeqNum;
-        this.elementCount = elementCount;
-
-        this.file = buildPath().toFile();
-        RandomAccessFile raf = new RandomAccessFile(this.file, "rw");
-        this.channel = raf.getChannel();
-        this.buffer = this.channel.map(FileChannel.MapMode.READ_WRITE, 0, this.capacity);
-        raf.close();
-        this.buffer.load();
-
-        this.buffer.position(0);
-        this.version = this.buffer.get();
-        this.head = 1;
-
-        if (this.elementCount > 0) {
-
-            // TODO: refactor the read logic below to DRY with the read() method.
-
-            // set head by skipping over all elements
-            for (int i = 0; i < this.elementCount; i++) {
-                if (this.head + SEQNUM_SIZE + LENGTH_SIZE > capacity) {
-                    throw new IOException(String.format("cannot read seqNum and length bytes past buffer capacity"));
-                }
-
-                long seqNum = this.buffer.getLong();
+        mapFile(STRICT_CAPACITY);
+        super.open(minSeqNum, elementCount);
+    }
 
-                if (i == 0 && seqNum != this.minSeqNum) {
-                    throw new IOException(String.format("first seqNum=%d is different than minSeqNum=%d", seqNum, this.minSeqNum));
-                }
+    // recover will overwrite/update/set this object minSeqNum, capacity and elementCount attributes
+    // to reflect what it recovered from the page
+    @Override
+    public void recover() throws IOException {
+        mapFile(!STRICT_CAPACITY);
+        super.recover();
+    }
 
-                this.offsetMap.add(head);
-                this.head += SEQNUM_SIZE;
+    // memory map data file to this.buffer and read initial version byte
+    // @param strictCapacity if true verify that data file size is same as confgured page capcity, if false update page capcity to actual file size
```

Comment:
oh you mean using `/** */` ?
I see I have been using `//` type of comments everywhere and this is not javadoc compatible, I will need to do a global cleanup.